___

<p style="text-align: center;"><img src="https://docs.google.com/uc?id=1lY0Uj5R04yMY3-ZppPWxqCr5pvBLYPnV" class="img-fluid" alt="CLRSWY"></p>

___

# WELCOME!

Welcome to "RFM Customer Segmentation & Cohort Analysis Project". This is the first project of the Capstone Project Series, which consists of 3 different project that contain different scenarios.

This is a project which you will learn what is RFM? And how to apply RFM Analysis and Customer Segmentation using Clustering algorithms. Also you will improve your Data Cleaning, Data Visualization and Exploratory Data Analysis capabilities. On the other hand you will create Cohort and Conduct Cohort Analysis.

Before diving into the project, please take a look at the determines and project structure.

- **NOTE:** This Project assumes that you already know the basics of coding in Python and are familiar with the theory behind Clustering.



# #Determines

Using the [Online Retail dataset](https://archive.ics.uci.edu/ml/datasets/Online+Retail) from the UCI Machine Learning Repository for exploratory data analysis, ***Customer Segmentation***, ***RFM Analysis***, ***Clustering*** and ***Cohort Analysis***.

This is a transnational data set which contains all the transactions occurring between 01/12/2010 and 09/12/2011 for a UK-based and registered non-store online retail. The company mainly sells unique all-occasion gifts. Many customers of the company are wholesalers.

Feature Information:

**InvoiceNo**: Invoice number. *Nominal*, a 6-digit integral number uniquely assigned to each transaction. If this code starts with letter 'c', it indicates a cancellation.
<br>
**StockCode**: Product (item) code. *Nominal*, a 5-digit integral number uniquely assigned to each distinct product.
<br>
**Description**: Product (item) name. *Nominal*.
<br>
**Quantity**: The quantities of each product (item) per transaction. *Numeric*.
<br>
**InvoiceDate**: Invoice Date and time. *Numeric*, the day and time when each transaction was generated.
<br>
**UnitPrice**: Unit price. *Numeric*, Product price per unit in sterling.
<br>
**CustomerID**: Customer number. *Nominal*, a 5-digit integral number uniquely assigned to each customer.
<br>
**Country**: Country name. *Nominal*, the name of the country where each customer resides.


---


First of all, to observe the structure of the data and missing values, you can use exploratory data analysis and data visualization techniques.

You must do descriptive analysis. Because you must understand the relationship of the features to each other and clear the noise and missing values in the data. After that, the data set will be ready for RFM analysis.

Before starting the RFM Analysis, you will be asked to do some analysis regarding the distribution of *Orders*, *Customers* and *Countries*. These analyzes will help the company develop its sales policies and contribute to the correct use of resources.

You will notice that the UK not only has the most sales revenue, but also the most customers. So you will continue to analyze only UK transactions in the next RFM Analysis, Customer Segmentation and Clustering topics.

Next, you will begin RFM Analysis, a customer segmentation technique based on customers' past purchasing behavior.

By using RFM Analysis, you can enable companies to develop different approaches to different customer segments so that they can get to know their customers better, observe trends better, and increase customer retention and sales revenues.

You will calculate the Recency, Frequency and Monetary values of the customers in the RFM Analysis you will make using the data consisting of UK transactions. Ultimately, you have to create an RFM table containing these values.

In the Customer Segmentation section, you will create an RFM Segmentation Table where you segment your customers by using the RFM table. For example, you can label the best customer as "Big Spenders" and the lost customer as "Lost Customer".

We will segment the customers ourselves based on their recency, frequency, and monetary values. But can an **unsupervised learning** model do this better for us? You will use the K-Means algorithm (any algorithm) to find the answer to this question. Then you will compare the classification made by the algorithm with the classification you have made yourself.

Before applying Clustering, you should do data pre-processing. In this context, it will be useful to examine feature correlations and distributions.

After the above operations, you will have made cluster estimation with clustering algorithms. You should visualize the cluster distribution by using a scatter plot. You can observe the properties of the resulting clusters with the help of the boxplot. Thus you will be able to tag clusters and interpret results.

Finally, you will do Cohort Analysis with the data you used at the beginning, regardless of the analysis you have done before. Cohort analysis is a subset of behavioral analytics that takes the user data and breaks them into related groups for analysis. This analysis can further be used to do customer segmentation and track metrics like retention, churn, and lifetime value.


## Our notes

- we can use any unsupervised machine learning algorithm
- RFM is for non-technical stakeholders e.g. used in marketing 
- RFM Analysis is used to understand and segment customers based on their buying behaviour
- Cohort analysis is similar to segmentation based on some metrics
- https://medium.com/@ugursavci/customer-segmentation-using-rfm-analysis-in-python-218a3255f714 (later reading)

# #Project Structures

- Data Cleaning & Exploratory Data Analysis
- RFM Analysis
- Customer Segmentation
- Applying Clustering
- Create Cohort and Conduct Cohort Analysis

# #Tasks

#### 1. Data Cleaning & Exploratory Data Analysis

- Import Modules, Load Data & Data Review
- Follow the Steps Below

    *i. Take a look at relationships between InvoiceNo, Quantity and UnitPrice columns.*
    
    *ii. What does the letter "C" in the invoiceno column mean?*
    
    *iii. Handling Missing Values*
    
    *iv. Clean the Data from the Noise and Missing Values*
    
    *v. Explore the Orders*
    
    *vi. Explore Customers by Country*
    
    *vii. Explore the UK Market*
    
#### 2. RFM Analysis

- Follow the steps below

   *i. Import Libraries*
   
   *ii. Review "df_uk" DataFrame (the df_uk what you create at the end of the Task 1)*
   
   *iii. Calculate Recency*
   
   *iv. Calculate Frequency*
   
   *v. Calculate Monetary Values*
   
   *vi. Create RFM Table*

#### 3. Customer Segmentation with RFM Scores
- Calculate RFM Scoring

    *i. Creating the RFM Segmentation Table*

- Plot RFM Segments

#### 4. Applying Clustering
- Data Pre-Processing and Exploring

    *i. Define and Plot Feature Correlations*

    *ii. Visualize Feature Distributions*

    *iii. Data Normalization*

- Clustering Implementation

    *i. Define Optimal Cluster Number (K) by using "Elbow Method" and "Silhouette Analysis"*

    *ii. Visualize the Clusters*

    *iii. Assign the label*

    *iv. Conclusion*

#### 5. Create Cohort and Conduct Cohort Analysis
- Future Engineering

    *i. Extract the Month of the Purchase*

    *ii. Calculating time offset in Months i.e. Cohort Index*

- Create 1st Cohort: User Number & Retention Rate

    *i. Pivot Cohort and Cohort Retention*

    *ii. Visualize analysis of cohort 1 using seaborn and matplotlib*

- Create 2nd Cohort: Average Quantity Sold

    *i. Pivot Cohort and Cohort Retention*

    *ii. Visualize analysis of cohort 2 using seaborn and matplotlib*

- Create 3rd Cohort: Average Sales

    *i. Pivot Cohort and Cohort Retention*

    *ii. Visualize analysis of cohort 3 using seaborn and matplotlib*
    
- **Note: There may be sub-tasks associated with each task, you will see them in order during the course of the work.**


# 1. Data Cleaning & Exploratory Data Analysis

## Import Modules, Load Data & Data Review

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import datetime
%matplotlib inline

import warnings
warnings.filterwarnings('ignore')

In [2]:
df = pd.read_excel('Online Retail.xlsx')
df.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[ns]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  object        
dtypes: datetime64[ns](1), float64(2), int64(1), object(4)
memory usage: 33.1+ MB


In [4]:
df.describe().T

count          mean          std       min       25%       50%  \
Quantity    541909.0      9.552250   218.081158 -80995.00      1.00      3.00   
UnitPrice   541909.0      4.611114    96.759853 -11062.06      1.25      2.08   
CustomerID  406829.0  15287.690570  1713.600303  12346.00  13953.00  15152.00   

                 75%      max  
Quantity       10.00  80995.0  
UnitPrice       4.13  38970.0  
CustomerID  16791.00  18287.0

Why do we have negative values???!

In [5]:
df.shape

(541909, 8)

we have 8 columns and 541909 rows

In [6]:
# visualize missing values 
details = df.isnull().sum().sort_values(ascending=False)
display(details)

CustomerID     135080
Description      1454
InvoiceNo           0
StockCode           0
Quantity            0
InvoiceDate         0
UnitPrice           0
Country             0
dtype: int64

In [7]:
541909 - 135080

406829

### checking duplactes before anf after

In [8]:
df.duplicated().sum()

5268

In [9]:
df.drop_duplicates(inplace=True)

In [10]:
details = df.isnull().sum().sort_values(ascending=False)
display(details)

CustomerID     135037
Description      1454
InvoiceNo           0
StockCode           0
Quantity            0
InvoiceDate         0
UnitPrice           0
Country             0
dtype: int64

### i. Take a look at relationships between InvoiceNo, Quantity and UnitPrice columns.

We see that there are negative values in the Quantity and UnitPrice columns. These are possibly canceled and returned orders. Let's check it out.

In [11]:
# display negative values in quantity column
df[df['Quantity'] < 0].shape[0]

10587

In [12]:
# display negative values in UnitPrice column
df[df['UnitPrice'] <0].shape[0]

2

Add comment later

### ii. What does the letter "C" in the InvoiceNo column mean?

hint: If the invoice number starts with the letter "C", it means the order was cancelled. Or those who abandon their order.

In [13]:
df['InvoiceNo'].str[0].value_counts()

C    9251
A       3
Name: InvoiceNo, dtype: int64

C stands for cancel 

In [14]:
df[df['InvoiceNo'].str[0] == 'A']

InvoiceNo StockCode      Description  Quantity         InvoiceDate  \
299982   A563185         B  Adjust bad debt         1 2011-08-12 14:50:00   
299983   A563186         B  Adjust bad debt         1 2011-08-12 14:51:00   
299984   A563187         B  Adjust bad debt         1 2011-08-12 14:52:00   

        UnitPrice  CustomerID         Country  
299982   11062.06         NaN  United Kingdom  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

In [15]:
(df['Description'] == 'Adjust bad debt').sum()

3

We think that A stands for "Adjust bad debt", appearantly the the transaction failed <br>
and for that reason there is no CustomerID and the unitprice is negative as well as the stockCode is B

In [16]:
df[df['InvoiceNo'].str[0] == 'C'][:10]

InvoiceNo StockCode                        Description  Quantity  \
141   C536379         D                           Discount        -1   
154   C536383    35004C    SET OF 3 COLOURED  FLYING DUCKS        -1   
235   C536391     22556     PLASTERS IN TIN CIRCUS PARADE        -12   
236   C536391     21984   PACK OF 12 PINK PAISLEY TISSUES        -24   
237   C536391     21983   PACK OF 12 BLUE PAISLEY TISSUES        -24   
238   C536391     21980  PACK OF 12 RED RETROSPOT TISSUES        -24   
239   C536391     21484        CHICK GREY HOT WATER BOTTLE       -12   
240   C536391     22557   PLASTERS IN TIN VINTAGE PAISLEY        -12   
241   C536391     22553             PLASTERS IN TIN SKULLS       -24   
939   C536506     22960           JAM MAKING SET WITH JARS        -6   

            InvoiceDate  UnitPrice  CustomerID         Country  
141 2010-12-01 09:41:00      27.50     14527.0  United Kingdom  
154 2010-12-01 09:49:00       4.65     15311.0  United Kingdom  
235 2010-12-01 10:24:00       1.65     17548.0  United Kingdom  
236 2010-12-01 10:24:00       0.29     17548.0  United Kingdom  
237 2010-12-01 10:24:00       0.29     17548.0  United Kingdom  
238 2010-12-01 10:24:00       0.29     17548.0  United Kingdom  
239 2010-12-01 10:24:00       3.45     17548.0  United Kingdom  
240 2010-12-01 10:24:00       1.65     17548.0  United Kingdom  
241 2010-12-01 10:24:00       1.65     17548.0  United Kingdom  
939 2010-12-01 12:38:00       4.25     17897.0  United Kingdom

Since the orders are cancelled the quantity is recorded in negative values

In [17]:
df.head(2)

InvoiceNo StockCode                         Description  Quantity  \
0    536365    85123A  WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                 WHITE METAL LANTERN         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

add comment here

In [ ]:
df[df['InvoiceNo'].str.startswith('C')].groupby(['InvoiceNo', 'Quantity'])

In [18]:
df.groupby('CustomerID')['InvoiceNo'].nunique().sort_values(ascending=False).head(10)

CustomerID
14911.0    248
12748.0    224
17841.0    169
14606.0    128
13089.0    118
15311.0    118
12971.0     89
14527.0     86
13408.0     81
14646.0     77
Name: InvoiceNo, dtype: int64

The number of items that each customer has purchace

When we filter canceled orders by Quantity > 0 or filter non-canceled orders by Quantity <0 nothing returns, this confirms that negative values mean the order was canceled. <br> So lets find out how many orders were cancelled?

In [20]:
df[df.Quantity > 0].shape[0]

526054

In [21]:
df[df.Quantity < 0].shape[0]

10587

In [22]:
df[df['InvoiceNo'].str[0] == 'C'].groupby('InvoiceNo').count().shape[0]

3836

In [34]:
df[df['Quantity'] < 0].groupby(df['InvoiceNo'].str[0] == 'C').sum()

Quantity  UnitPrice   CustomerID
InvoiceNo                                  
False       -206957       0.00          0.0
True        -275560  449325.05  132992637.0

In [42]:
df[(df['InvoiceNo'].str[0] == 'C') & (df['Quantity'] > 0) ].count()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

when the order is cancelled the quantity is negative

In [43]:
df[(df['InvoiceNo'].str[0] == 'C') & (df['Quantity'] < 0) ].count()

InvoiceNo      9251
StockCode      9251
Description    9251
Quantity       9251
InvoiceDate    9251
UnitPrice      9251
CustomerID     8872
Country        9251
dtype: int64

In [44]:
df[(df['InvoiceNo'].str[0] != 'C') & (df['Quantity'] < 0) ].count()

InvoiceNo      1336
StockCode      1336
Description     474
Quantity       1336
InvoiceDate    1336
UnitPrice      1336
CustomerID        0
Country        1336
dtype: int64

overall we have 10587 negative Quantity<br>
there is 9251 cancelled and negative <br>
there is 1336 not cancelled but negative (possible outliers?)

### So lets find out how many orders were cancelled?

In [49]:
df['InvoiceNo'].value_counts().unique()

array([1114,  749,  731,  721,  705,  687,  676,  675,  662,  652,  650,
        645,  635,  629,  622,  620,  606,  601,  597,  593,  592,  591,
        585,  578,  572,  568,  561,  552,  544,  542,  541,  536,  534,
        533,  529,  527,  526,  520,  518,  515,  512,  503,  502,  494,
        490,  487,  486,  482,  474,  464,  462,  460,  458,  455,  450,
        444,  442,  438,  436,  435,  434,  429,  421,  418,  416,  415,
        411,  410,  406,  404,  403,  402,  392,  388,  384,  379,  378,
        377,  371,  364,  360,  358,  355,  353,  352,  350,  347,  346,
        344,  343,  340,  335,  334,  333,  332,  320,  319,  315,  310,
        309,  301,  300,  298,  297,  295,  293,  292,  290,  289,  288,
        287,  285,  284,  283,  282,  281,  280,  275,  274,  271,  267,
        266,  264,  263,  262,  261,  259,  257,  256,  255,  254,  253,
        250,  249,  247,  246,  245,  244,  243,  242,  241,  239,  238,
        237,  236,  235,  234,  233,  230,  229,  2

In [52]:
df.groupby('InvoiceNo').count()

StockCode  Description  Quantity  InvoiceDate  UnitPrice  \
InvoiceNo                                                             
536365             7            7         7            7          7   
536366             2            2         2            2          2   
536367            12           12        12           12         12   
536368             4            4         4            4          4   
536369             1            1         1            1          1   
...              ...          ...       ...          ...        ...   
C581484            1            1         1            1          1   
C581490            2            2         2            2          2   
C581499            1            1         1            1          1   
C581568            1            1         1            1          1   
C581569            2            2         2            2          2   

           CustomerID  Country  
InvoiceNo                       
536365              7        7  
536366              2        2  
536367             12       12  
536368              4        4  
536369              1        1  
...               ...      ...  
C581484             1        1  
C581490             2        2  
C581499             1        1  
C581568             1        1  
C581569             2        2  

[25900 rows x 7 columns]

In [76]:
df['InvoiceNo'].unique().shape[0]

25900

there is around 26 K unique orders recorded in the dataset

In [65]:
df[df['InvoiceNo'].str[0] == 'C'].groupby('InvoiceNo').count().shape[0]

3836

In [72]:
df[df['InvoiceNo'].str[0] == 'C']['InvoiceNo'].unique().shape[0]

3836

we have around 3800 cancelled orders (unique)

In [73]:
df[df['InvoiceNo'].str[0] == 'C']['InvoiceNo'].shape[0]

9251

In [74]:
9251 - 3836

5415

duplacated cancelled orders

#### Looking deeper into why these orders were cancelled may prevent future cancellations. <br> Now let's find out what a negative UnitPrice means.


In [79]:
df[(df['UnitPrice'] < 0) & (df['Quantity'] < 0) ].count()

InvoiceNo      0
StockCode      0
Description    0
Quantity       0
InvoiceDate    0
UnitPrice      0
CustomerID     0
Country        0
dtype: int64

In [82]:
df[df['UnitPrice'] < 0]

InvoiceNo StockCode      Description  Quantity         InvoiceDate  \
299983   A563186         B  Adjust bad debt         1 2011-08-12 14:51:00   
299984   A563187         B  Adjust bad debt         1 2011-08-12 14:52:00   

        UnitPrice  CustomerID         Country  
299983  -11062.06         NaN  United Kingdom  
299984  -11062.06         NaN  United Kingdom

we already investegated that :)

### iii. Handling Missing Values

Since the customer ID's are missing, lets assume these orders were not made by the customers already in the data set because those customers already have ID's.

We also don't want to assign these orders to those customers because this would alter the insights we draw from the data.


In [ ]:
df.dropna(inplace= True, index= 0)

### iv. Clean the Data from the Noise and Missing Values

### v. Explore the Orders


1. Find the unique number of InvoiceNo  per customer

2. What's the average number of unqiue items per order or per customer?

3. Let's see how this compares to the number of unique products per customer.

### vi. Explore Customers by Country

1. What's the total revenue per country?

2. Visualize number of customer per country

3. Visualize total cost per country

#### The UK not only has the most sales revenue, but also the most customers. Since the majority of this data set contains orders from the UK, we can explore the UK market further by finding out what products the customers buy together and any other buying behaviors to improve our sales and targeting strategy.

### vii. Explore the UK Market


1. Create df_uk DataFrame

2. What are the most popular products that are bought in the UK?

### We will continue analyzing the UK transactions with customer segmentation.

# 2. RFM Analysis

In the age of the internet and e-commerce, companies that do not expand their businesses online or utilize digital tools to reach their customers will run into issues like scalability and a lack of digital precsence. An important marketing strategy e-commerce businesses use for analyzing and predicting customer value is customer segmentation. Customer data is used to sort customers into group based on their behaviors and preferences.

**[RFM](https://www.putler.com/rfm-analysis/) (Recency, Frequency, Monetary) Analysis** is a customer segmentation technique for analyzing customer value based on past buying behavior. RFM analysis was first used by the direct mail industry more than four decades ago, yet it is still an effective way to optimize your marketing.
<br>
<br>
Our goal in this Notebook is to cluster the customers in our data set to:
 - Recognize who are our most valuable customers
 - Increase revenue
 - Increase customer retention
 - Learn more about the trends and behaviors of our customers
 - Define customers that are at risk

We will start with **RFM Analysis** and then compliment our findings with using **Clustering Algorithms.**

- RECENCY (R): Time since last purchase
- FREQUENCY (F): Total number of purchases
- MONETARY VALUE (M): Total monetary value




Benefits of RFM Analysis

- Increased customer retention
- Increased response rate
- Increased conversion rate
- Increased revenue

RFM Analysis answers the following questions:
 - Who are our best customers?
 - Who has the potential to be converted into more profitable customers?
 - Which customers do we need to retain?
 - Which group of customers is most likely to respond to our marketing campaign?


### i. Import Libraries

### ii. Review df_uk DataFrame

### iii. Recency: Days since last purchase
To calculate the recency values, follow these steps in order:

1. To calculate recency, we need to choose a date as a point of reference to evaluate how many days ago was the customer's last purchase.
2. Create a new column called Date which contains the invoice date without the timestamp
3. Group by CustomerID and check the last date of purchase
4. Calculate the days since last purchase
5. Drop Last_Purchase_Date since we don't need it anymore
6. Plot RFM distributions

1. Choose a date as a point of reference to evaluate how many days ago was the customer's last purchase.

2. Create a new column called Date which contains the invoice date without the timestamp

3. Group by CustomerID and check the last date of purchase

4. Calculate the days since last purchase

5. Drop Last_Purchase_Date since we don't need it anymore

6. Plot RFM distributions

### iv. Frequency: Number of purchases

To calculate how many times a customer purchased something, we need to count how many invoices each customer has. To calculate the frequency values, follow these steps in order:

1. Make a copy of df_uk and drop duplicates

2. Calculate the frequency of purchases

3. Plot RFM distributions

### v. Monetary: Total amount of money spent

The monetary value is calculated by adding together the cost of the customers' purchases.


1. Calculate sum total cost by customers and named "Monetary"

2. Plot RFM distributions

### vi. Create RFM Table
Merge the recency, frequency and motetary dataframes

# 3. Customer Segmentation with RFM Scores

Businesses have this ever-lasting urge to understand their customers. The better you understand the customer, the better you serve them, and the higher the financial gain you receive from that customer. Since the dawn of trade, this process of understanding customers for a strategic gain has been there practiced and this task is known majorly as [Customer Segmentation](https://clevertap.com/blog/rfm-analysis/).
Well as the name suggests, Customer Segmentation could segment customers according to their precise needs. Some of the common ways of segmenting customers are based on their Recency-Frequency-Monatory values, their demographics like gender, region, country, etc, and some of their business-crafted scores. You will use Recency-Frequency-Monatory values for this case.

In this section, you will create an RFM Segmentation Table where you segment your customers by using the RFM table. For example, you can label the best customer as "Big Spenders" and the lost customer as "Lost Customer".

## Calculate RFM Scoring

The simplest way to create customer segments from an RFM model is by using **Quartiles**. We will assign a score from 1 to 4 to each category (Recency, Frequency, and Monetary) with 4 being the highest/best value. The final RFM score is calculated by combining all RFM values. For Customer Segmentation, you will use the df_rfm data set resulting from the RFM analysis.
<br>
<br>
**Note**: Data can be assigned into more groups for better granularity, but we will use 4 in this case.

1. Divide the df_rfm into quarters

### i. Creating the RFM Segmentation Table


1. Create two functions, one for Recency and one for Frequency and Monetary. For Recency, customers in the first quarter should be scored as 4, this represents the highest Recency value. Conversely, for Frequency and Monetary, customers in the last quarter should be scored as 4, representing the highest Frequency and Monetary values.

2. Score customers from 1 to 4 by applying the functions you have created. Also create separate score column for each value.

3. Now that scored each customer, you'll combine the scores for segmentation.

4. Define rfm_level function that tags customers by using RFM_Scrores and Create a new variable RFM_Level

5. Calculate average values for each RFM_Level, and return a size of each segment

## Plot RFM Segments

1. Create your plot and resize it.

Using customer segmentation categories found [here](http://www.blastam.com/blog/rfm-analysis-boosts-sales). We can formulate different marketing strategies and approaches for customer engagement for each type of customer.

Note: The author in the article scores 1 as the highest and 4 as the lowest

2. How many customers do we have in each segment?

# 3. Applying K-Means Clustering

Now that we have our customers segmented into different categories, we can gain further insight into customer behavior by using unsupervised models in conjuction with out RFM model.
Unsupervised Learning has also been widely used for market segmentation and has the advantage of being simple to implement. Possible algorithms include **K-means Clustering**, **Aglomerative Clustering**, and **Gaussian Mixtute**.

## Data Pre-Processing and Exploring

### i. Define and Plot Feature Correlations

Create Heatmap and evaluate the results

### ii. Visualize Feature Distributions

To get a better understanding of the dataset, you can costruct a scatter matrix of each of the three features in the RFM data.

### iii. Data Normalization

1. Logarithmic Tranformation can be used to normalize the values in a column.
2. Plot normalized data with scatter matrix or pairplot. Also evaluate results.

## Clustering Implementation

Try to apply different clustering algoritms to gather clusters like you achieved in manual RFM analysis.

### i. Define the Optimal Number of Clusters

### ii. Model Fitting

### iii. Visualize the Clusters

2. Visualize Cluster Id vs Recency, Cluster Id vs Frequency and Cluster Id vs Monetary using Box plot. Also evaluate the results.

### iv. Assign the Label

### v. Conclusion

Discuss your final results. Compare your own labels from the RFM Customer Segmentation with the labels found by Unsupervised Algorithms.

How we want to continue this analysis depends on how the business plans to use the results and the level of granularity the business stakeholders want to see in the clusters. We can also ask what range of customer behavior from high to low value customers are the stakeholders interested in exploring. From those answers, various methods of clustering can be used and applied on RFM variable or directly on the transaction data set.

**Conclusion**

* Write down your conclusions about clustering

# 5. Create Cohort & Conduct Cohort Analysis
[Cohort Analysis](https://medium.com/swlh/cohort-analysis-using-python-and-pandas-d2a60f4d0a4d) is specifically useful in analyzing user growth patterns for products. In terms of a product, a cohort can be a group of people with the same sign-up date, the same usage starts month/date, or the same traffic source.
Cohort analysis is an analytics method by which these groups can be tracked over time for finding key insights. This analysis can further be used to do customer segmentation and track metrics like retention, churn, and lifetime value.

For e-commerce organizations, cohort analysis is a unique opportunity to find out which clients are the most valuable to their business. by performing Cohort analysis you can get the following answers to the following questions:

- How much effective was a marketing campaign held in a particular time period?
- Did the strategy employ to improve the conversion rates of Customers worked?
- Should I focus more on retention rather than acquiring new customers?
- Are my customer nurturing strategies effective?
- Which marketing channels bring me the best results?
- Is there a seasonality pattern in Customer behavior?
- Along with various performance measures/metrics for your organization.

Since we will be performing Cohort Analysis based on transaction records of customers, the columns we will be dealing with mainly:
- Invoice Data
- CustomerID
- Price
- Quantity

The following steps will performed to generate the Cohort Chart of Retention Rate:
- Month Extraction from InvioceDate column
- Assigning Cohort to Each Transaction
- Assigning Cohort Index to each transaction
- Calculating number of unique customers in each Group of (ChortDate,Index)
- Creating Cohort Table for Retention Rate
- Creating the Cohort Chart using the Cohort Table

The Detailed information about each step is given below:

## Future Engineering

### i. Extract the Month of the Purchase
First we will create a function, which takes any date and returns the formatted date with day value as 1st of the same month and Year.

Now we will use the function created above to convert all the invoice dates into respective month date format.

### ii. Calculating time offset in Months i.e. Cohort Index:
Calculating time offset for each transaction will allows us to report the metrics for each cohort in a comparable fashion.
First, you will create 4 variables that capture the integer value of years, months for Invoice and Cohort Date using the get_date_int() function which you'll create it below.

You will use this function to extract the integer values for Invoice as well as Cohort Date in 3 separate series for each of the two columns

Use the variables created above to calcualte the difference in days and store them in cohort Index column.

## Create 1st Cohort: User number & Retention Rate

### i. Pivot Cohort and Cohort Retention

### ii. Visualize analysis of cohort 1 using seaborn and matplotlib modules

## Create the 2nd Cohort: Average Quantity Sold

### i. Pivot Cohort and Cohort Retention

### ii. Visualize analysis of cohort 2 using seaborn and matplotlib modules

## Create the 3rd Cohort: Average Sales


### i. Pivot Cohort and Cohort Retention

### ii. Visualize analysis of cohort 3 using seaborn and matplotlib modules

For e-commerce organisations, cohort analysis is a unique opportunity to find out which clients are the most valuable to their business. by performing Cohort analysis you can get answers to following questions:

- How much effective was a marketing campaign held in a particular time period?
- Did the strategy employed to improve the conversion rates of Customers worked?
- Should I focus more on retention rather than acquiring new customers?
- Are my customer nurturing strategies effective?
- Which marketing channels bring me the best results?
- Is there a seasoanlity pattern in Customer behahiour?

___

<p style="text-align: center;"><img src="https://docs.google.com/uc?id=1lY0Uj5R04yMY3-ZppPWxqCr5pvBLYPnV" class="img-fluid" alt="CLRSWY"></p>

___